### PR TITLE
Decoder refactor

### PR DESCRIPTION
**Pull request by jindrahelcl**
Refactor of the decoder.

Main change is separation of input and training placeholders and a partial conversion from lists to tensors.

- Input placeholder is the go_symbols tensor, filled with a batch of starting symbols (this cannot be inferred since we still want to have the option for no encoder and we need to create the initial state and this needs, in turn, the batch size)
- Training placeholder is the decoder inputs and padding. Note these training inputs still contain their own go symbols, which is a subject for a further refactor
- Decoding function works with tensors, but creates lists. That's why the "partial" in the "partial conversion".

- `decoding_function` is simplified and `decode_step` is derived from it. Scopes are now a bit of a mess, but `assert`s are added to ensure nothing is silently copied (I think it cannot be copied anyway, but let's keep the assertions there)

### CODE REVIEW

File: neuralmonkey/decoders/decoder.py [hunk 1]
```diff
@@ -185,60 +221,72 @@ def _encoder_projection(self, encoded_states):
...
+            self.embedding_matrix = tf.get_variable(
+                "word_embeddings", [self.vocabulary_size, self.embedding_size],
+                initializer=tf.random_normal_initializer(stddev=0.01))
 
-    def _training_placeholders(self):
-        """Defines data placeholders for training the decoder"""
 
-        inputs = [tf.placeholder(tf.int64, [None], name="decoder_{}".format(i))
-                  for i in range(self.max_output + 2)]
+    def _embed(self, inputs):
```
Comment (jlibovicky) [inline neuralmonkey/decoders/decoder.py]: To bych přejmenoval na `_embed_and_dropout`, ať to křičí z názvu i bez docstringu.

File: neuralmonkey/runners/rnn_runner.py [hunk 2]
```diff
@@ -241,19 +241,19 @@ def next_to_execute(self) -> NextExecute:
...
 
-        to_run = {'logprobs': self._decoder.train_logprobs[self._time_step]}
+        to_run = {'logprobs': self._decoder.runtime_logprobs[self._time_step]}
 
         self._current_beam_batch = self._to_exapand.pop()
 
         if self._current_beam_batch is not None:
             additional_feed_dict = {t: index for t, index in zip(
-                self._decoder.train_inputs[1:],
+                tf.unpack(self._decoder.train_inputs[1:]),
```
Comment (jlibovicky) [inline neuralmonkey/runners/rnn_runner.py]: Když se tady feeduje do `train_inputs`, tak to ale přece vůbec neovlivní ty `runtime_logprobs`, ty závísí jenom na loop function, takže nám pak vždycky dají distribuci, která odpovídá greedy dekódování.

File: neuralmonkey/decoders/decoder.py [hunk 3]
```diff
@@ -62,83 +77,80 @@ def __init__(self, encoders, vocabulary, data_id, name, **kwargs):
...
-            "learning_step", [], initializer=tf.constant_initializer(0),
-            trainable=False)
-
-        if self.project_encoder_outputs or len(self.encoders) == 0:
-            self.rnn_size = kwargs.get("rnn_size", 200)
-        else:
+        if len(self.encoders) > 0 and not self.project_encoder_outputs:
             if "rnn_size" in kwargs:
                 log("Warning: rnn_size attribute will not be used "
                     "without encoder projection!", color="red")
```
Comment (jlibovicky) [inline neuralmonkey/decoders/decoder.py]: Tady by se mohla ta hláška rozšířit o tom, jaká se nastavila velikost a proč.

File: neuralmonkey/decoders/decoder.py [hunk 4]
```diff
@@ -308,88 +337,91 @@ def _logit_function(self, rnn_output):
...
+            output_logits.append(logit)
+            rnn_outputs.append(output)
+            rnn_states.append(state)
 
         return rnn_outputs, rnn_states, output_logits
 
 
-    def _init_summaries(self):
-        """Initialize the summaries of the decoder
+    def _visualize_alignments(self):
```
Comment (jlibovicky) [inline neuralmonkey/decoders/decoder.py]: Já bych tomu neříkal alignments, Ondra se chystá napsat použití předtrénovaného alignmentu, tak bych radši pořádně všude odlišoval, co je alignment a co je attention, aby někdy v budoucnu nedošlo k nějakým zmatkům.

File: neuralmonkey/runners/rnn_runner.py (hunk 2 above)
Comment (jindrahelcl) [inline neuralmonkey/runners/rnn_runner.py]: Tak co se tady má feedovat? Nic? Nebo nějakej tenzor? RNN Runner by mohl z dekodéru používat jen tu funkci decoding_step. Ale do přepisování RNN runneru se hrnout nechci,  protože tomu kódu moc nerozumim. Ale chtělo by to dát sem něco uspokojivějšího, to jo..

File: neuralmonkey/decoders/decoder.py (hunk 3 above)
Comment (jindrahelcl) [inline neuralmonkey/decoders/decoder.py]: ✓

File: neuralmonkey/decoders/decoder.py (hunk 1 above)
Comment (jindrahelcl) [inline neuralmonkey/decoders/decoder.py]: ✓

File: neuralmonkey/decoders/decoder.py (hunk 4 above)
Comment (jindrahelcl) [inline neuralmonkey/decoders/decoder.py]: ✓

File: neuralmonkey/runners/rnn_runner.py (hunk 2 above)
Comment (jlibovicky) [inline neuralmonkey/runners/rnn_runner.py]: Pokud tam skutečně chceme používat `runtime_logprobs`, tak je potřeba feedovat tenzor, co vzniká jako `previous_word` v loop function.

**Comment by jindrahelcl** [comment]
Ok this needs to be merged ASAP because all of you are commiting changes everywhere and the conflicts just keep on multiplying.. Rebasing this is becoming a pure hell.

**Comment by tomasmcz** [comment]
Is it necessary to merge all those 41 commits at once? Not only is this hard to rebase, it's also hard to review. Are there any parts that make sense on their own? Maybe they could be separated (with interactive rebase and/or `git filter-branch`), reviewed faster and merged. Or maybe that's not worth it and we should just merge this and try to keep the PRs shorter in the future, I don't know.

**Comment by jindrahelcl** [comment]
Yes. It's completely new decoder. It's already rebased (that's why the frustrated message). Moreover, @jlibovicky has done a review somewhere after 20 commits.

Anyway, I must still finish some things. Every time I commit even a minor change, I'm running an experiment to confirm that the decoder still works, so it takes time.

**Comment by jindrahelcl** [comment]
Ale je fakt, že do jistý míry to rozsekat jde, tak to vyzkoušim a zkusim se o to pokusit.

**Comment by jindrahelcl** [comment]
Doběhlo to v pořádku, BLEU 30..

**Comment by tomasmcz** [comment]
Opravte testy a pak to mergujte.
